### PR TITLE
Add support for downloading images from the Figma api marked as exportable

### PR DESCRIPTION
**Pull request by sHoskens**
I'll be trying to add some valid tests too if I find any spare time, but this can be merged in already.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2,12 +2,14 @@ const fs = require('fs');
...
-const { prettyJSON, writeFile } = require('./lib/utils');
+const { prettyJSON, writeFile, downloadFile } = require('./lib/utils');
 const mapToStyleDictionaryTokens = require('./lib/styleDictionary');
 
+const ASSETS_DIR = 'assets';
+
 module.exports = async (args, flags) => {
   if (flags.version) return pkg.version;
-  if (args.length <= 0) throw new Error('No file input passed after npm start');
+  if (args.length <= 0 && !flags.token) throw new Error('No file input or token passed after npm start');
```
Comment (thibmaek) [inline index.js]: We can not require a token since Sketch does not use a token, only Figma does.
Checking tokens should be done in the correct parser

**Review by thibmaek — CHANGES_REQUESTED** [review]

File: lib/utils.js [hunk 2]
```diff
@@ -142,6 +144,33 @@ module.exports = {
...
   },
 
+  downloadFile: (downloadDir, { fileUrl, fileName }) => {
+    const options = {
+      host: url.parse(fileUrl).host,
+      port: 80,
+      path: url.parse(fileUrl).pathname
+    };
+
+    const safeDownloadDir = `${downloadDir}${!downloadDir.endsWith('/') ? '/' : ''}`;
```
Comment (thibmaek) [inline lib/utils.js]: Please use path.join or other safe Node Path APIs to construct folder paths.

File: index.js [hunk 3]
```diff
@@ -50,6 +53,18 @@ module.exports = async (args, flags) => {
     await writeFile(`${flags.outputDir}/hubble-data.json`, prettyJSON(mapping));
   }
 
+  if (filesToDownload) {
+    const fullAssetsDir = `${flags.outputDir}/${ASSETS_DIR}/`;
```
Comment (thibmaek) [inline index.js]: Ending paths with a trailing slash is generally considered less clean when you are concatenating them in an unsafe way:

```js
const path = '/usr/local/'
const binPath = `${path}/bin`
console.log(binPath) // /usr/local//bin
```

File: lib/figma.js [hunk 4]
```diff
@@ -45,3 +45,100 @@ exports.findAllTokens = (data, tokenType) =>
...
+        if (setting.constraint && setting.constraint.type && setting.constraint.type === 'SCALE') {
+          scale = setting.constraint.value;
+        }
+
+        const existingElement = acc.find(c => c.format === format && c.scale === scale);
+        if (existingElement) {
+          existingElement.ids.push(component.id);
+        } else {
+          acc.push({
+            ids: [component.id],
```
Comment (thibmaek) [inline lib/figma.js:73]: Why would id need to be a tuple instead of just a string?

File: lib/figma.js [hunk 5]
```diff
@@ -45,3 +45,100 @@ exports.findAllTokens = (data, tokenType) =>
...
+    );
+
+    // Now create an array of objects containing the full desired filename and url.
+    return imageRequestResults.reduce((acc, result) => {
+      result.urls.forEach(({ id, url }) => {
+        const matchingFileData = exportableData.find(d => d.id === id);
+        acc.push({
+          id,
+          fileUrl: url,
+          fileName: `${matchingFileData.name}${result.suffix}.${result.format}`,
```
Comment (thibmaek) [inline lib/figma.js]: Ensure that filenames are always lowercase snake cased. This is required for assets exported with Hubble.

File: lib/parser/index.js [hunk 6]
```diff
@@ -60,6 +61,7 @@ const parseTokens = async (args, flags) => {
     try {
       const { token } = flags;
       if (!token) throw new Error('Please add a Figma API authorization token');
+      if (!file) throw new Error('Please add the unique id of the target Figma file');
```
Comment (thibmaek) [inline lib/parser/index.js]: This check is already performed in index.js:10

**Review by thibmaek — CHANGES_REQUESTED** [review]

**Comment by thibmaek** [comment]
Thinking that it would probably be best if we do _not_ export assets by default but enable it with a flag `--exportAssets|-e`

File: lib/figma.js (hunk 4 above)
Comment (sHoskens) [reply lib/figma.js:73]: Adding all image asset ids with the same format and scale to the same object, to limit the amount of Figma API requests necessary. The API call for images is not really intended to be used per asset, but per format, with an array of asset ids.

**Comment by thibmaek** [comment]
Looks good at first sight. Please add a case where we verify this works in the mocks. I would update generate_mocks.sh and check that the expected files are there.

**Review by thibmaek — APPROVED** [review]

**Pull request merged**